### PR TITLE
Fix for proxy-backend plugin when global-agent is enabled

### PR DESCRIPTION
**Pull request by bobalong79**
#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

#### Summary

When [global-agent](https://github.com/gajus/global-agent) is enabled the proxy-backend plugin generates a fatal error when sending the proxy request:  `Error [ERR_HTTP_HEADERS_SENT]: Cannot remove headers after they are sent to the client`. This is because the [`onProxyReq`](https://github.com/backstage/backstage/blob/master/plugins/proxy-backend/src/service/router.ts#L108) event fires _after_ the agent has already sent the headers to the target. This has been reported several times in the following issues:

* https://github.com/chimurai/http-proxy-middleware/issues/472
* https://github.com/chimurai/http-proxy-middleware/issues/318
* https://github.com/http-party/node-http-proxy/issues/1292
* https://github.com/http-party/node-http-proxy/issues/1502
* https://github.com/http-party/node-http-proxy/issues/1478

Looking through these reports it's clear that the issue is in `http-proxy`, not `http-proxy-middleware` used by the proxy-backend plugin.

Note that the problem only occurs in the [header filtering for the proxy request](https://github.com/backstage/backstage/blob/49be39b92a7c2c76f1a8f709c165de20dcede400/plugins/proxy-backend/src/service/router.ts#L93-L116). The [header filtering for the client response](https://github.com/backstage/backstage/blob/49be39b92a7c2c76f1a8f709c165de20dcede400/plugins/proxy-backend/src/service/router.ts#L118-L138) is unaffected.

#### Reproduce steps

* Enable `global-agent` as described in [help-im-behind-a-corporate-proxy.md](https://github.com/backstage/backstage/blob/49be39b92a7c2c76f1a8f709c165de20dcede400/contrib/docs/tutorials/help-im-behind-a-corporate-proxy.md)
* Add some proxy targets (I'm using the standard SonarQube and Jira plugin config)
* Send some requests to to the proxy and observe the `ERR_HTTP_HEADERS_SENT` crash
* Ensure that your requests include headers not in the [allow list](https://github.com/backstage/backstage/blob/49be39b92a7c2c76f1a8f709c165de20dcede400/plugins/proxy-backend/src/service/router.ts#L28-L47)

There are reports about seeing this when **any** [`http.Agent`](https://nodejs.org/api/http.html#http_class_http_agent) is registered on the proxy connection but I haven't tested with any other agents than the ones automatically set by `global-agent` ([`HttpProxyAgent`](https://github.com/gajus/global-agent/blob/f89883c39e7053bcc0d31b12c9c61d04961779df/src/classes/HttpProxyAgent.js) and [`HttpsProxyAgent`](https://github.com/gajus/global-agent/blob/f89883c39e7053bcc0d31b12c9c61d04961779df/src/classes/HttpsProxyAgent.js))

#### Workaround

My workaround is to use `http-proxy-middleware`'s [custom filter mechanism](https://github.com/chimurai/http-proxy-middleware/blob/3120f701f9b91ca0d6f32e1de5f1aaaa65520011/recipes/context-matching.md#custom-filtering) to alter the headers prior to the proxy request being set up. This filter mechanism was designed to allow you to implement custom matching logic, but it provides access to the `req: IncomingRequest` object [before the proxy request is created](https://github.com/chimurai/http-proxy-middleware/blob/3120f701f9b91ca0d6f32e1de5f1aaaa65520011/src/http-proxy-middleware.ts#L51), making it a convenient way to manipulate the headers before the `http-proxy` code is run.

As backstage [already uses the custom filter](https://github.com/backstage/backstage/blob/49be39b92a7c2c76f1a8f709c165de20dcede400/plugins/proxy-backend/src/service/router.ts#L88-L91) to reject proxy requests that don't match the allowed METHODs, I simply moved the existing header filtering code from `onProxyReq` into that function. 

Looking at what others have done, this approach is similar to the workarounds I've seen, although in most cases people did the header manipulation in a separate upstream middleware.

#### Testing

I've tested this with and without `global-agent` but I haven't added any regression tests as I wasn't sure how to register the agent and supporting forward proxy. Ideally we'd run the header tests twice, once without any agent being used and once with an agent. Setting up a mock forward proxy and using an agent-registration approach similar to that of `global-agent` looks like it would be complicated and difficult, however if the theory that this happens with any agent is correct, then setting up the test might be more doable? Suggestions welcome, though personally I think it's ok proceed without agent tests in this case.

### CODE REVIEW

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4740-x3f2g7gyqnbhbzsgntaik1emahal6zpj.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/603ce8e065f817845dfd5e74

File: .changeset/nervous-dogs-pull.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-proxy-backend': minor
```
Comment (Rugvip) [inline .changeset/nervous-dogs-pull.md:2]: While on 0.x we're marking non-breaking changes as `patch` level, which I think this should be

**Review by Rugvip — COMMENTED** [review]
Makes sense, thank you for the thorough research and write-up! :grin:

Just a tweak to the changeset and I think we're good to go.

Regarding tests for global-agent, we could add a TODO or create an issue for it, but I also think it's alright to skip for now. One thing is perhaps to add the rationale to the new comment, explaining why we do the header modifications in the filter rather than in `onProxyReq`.

File: plugins/proxy-backend/src/service/router.ts [hunk 2]
```diff
@@ -104,15 +99,18 @@ export function buildMiddleware(
     ].map(h => h.toLocaleLowerCase()),
   );
 
-  // only forward the allowed headers in client->backend
-  fullConfig.onProxyReq = (proxyReq: http.ClientRequest) => {
```
Comment (dhenneke) [inline plugins/proxy-backend/src/service/router.ts:108]: You need to update the tests in [router.test.ts](https://github.com/backstage/backstage/blob/9d8515ba68470bd0eb0ee1b654156857e8bd52d7/plugins/proxy-backend/src/service/router.test.ts#L136) as they explicitly check if the `onProxyReq` is setup correctly and they now need to be updated to check if the `filter` function is setup correctly.

Comment (bobalong79) [reply plugins/proxy-backend/src/service/router.ts:108]: Thanks, am fixing the tests now. I think I'll need to separate the header processing into a different function that's called from the filter so I can test it.

**Comment by backstage-service** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4740-x3f2g7gyqnbhbzsgntaik1emahal6zpj.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/603ce8e065f817845dfd5e74

File: plugins/proxy-backend/src/service/router.ts (hunk 2 above)
Comment (bobalong79) [reply plugins/proxy-backend/src/service/router.ts:108]: Hi - I didn't separate out the filtering in the end, but I've updated the tests to check that the correct filtering is applied to `req.headers` after `filter()` is called

File: plugins/proxy-backend/src/service/router.test.ts [hunk 3]
```diff
@@ -223,7 +231,7 @@ describe('buildMiddleware', () => {
     } as Partial<http.IncomingMessage>;
 
     expect(config).toBeDefined();
-    expect(config.onProxyReq).toBeDefined();
+    expect(config.onProxyRes).toBeDefined();
```
Comment (dhenneke) [inline plugins/proxy-backend/src/service/router.test.ts:234]: 😱 oops. good catch 👍

**Review by dhenneke — APPROVED** [review]
Thanks!

**Review by Rugvip — APPROVED** [review]
Awesome, thanks! 👍  :shipit:

**Pull request merged**